plugin.rb

```diff
@@ -217,40 +194,28 @@ def can_vote?
       end
 
       def vote_count
-        if count = self.custom_fields[DiscourseVoting::VOTE_COUNT]
-          # we may have a weird array here, don't explode
-          # need to fix core to enforce types on fields
-          count.try(:to_i) || 0
-        else
-          0 if self.can_vote?
-        end
+        self.vote_counter&.counter.to_i
       end
 
-      def user_voted(user)
-        if user && user.custom_fields[DiscourseVoting::VOTES]
-          user.custom_fields[DiscourseVoting::VOTES].include? self.id
+      def user_voted?(user)
+        if self.current_user_voted
+          self.current_user_voted && self.current_user_voted > 0
         else
-          false
+          votes.map(&:user_id).include?(user.id)
         end
       end
 
       def update_vote_count
-        count =
-          UserCustomField.where("value = :value AND name IN (:keys)",
-                                value: id.to_s, keys: [DiscourseVoting::VOTES, DiscourseVoting::VOTES_ARCHIVE]).count
+        count = self.votes.count
 
-        custom_fields[DiscourseVoting::VOTE_COUNT] = count
-        save_custom_fields
+        counter = self.vote_counter || DiscourseVoting::VoteCounter.new(topic: self)
+        counter.update(counter: count)
       end
 
       def who_voted
         return nil unless SiteSetting.voting_show_who_voted
-
-        User.where("id in (
-        SELECT user_id FROM user_custom_fields WHERE name IN (:keys) AND value = :value
-      )", value: id.to_s, keys: [DiscourseVoting::VOTES, DiscourseVoting::VOTES_ARCHIVE])
+        self.votes.map(&:user)
```

Comment:
Just want to confirm that we're eager loading users here so that this doesn't cause an N+1. Previously we would only execute a single query.

Reply:
good question, I ensured that who voted is only available if you see specific topic (not on topics list)